---
output:
  md_document:
    variant: gfm
html_preview: false
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# BMplot: Base Modification Visualization

<!-- badges: start -->

<!-- badges: end -->

This package implements functions to visualize the profile of base modification in both DNA and RNA.

## Introduction

Biology relies heavily on epigenetics as a method of altering genetic information without altering nucleic acids.
  
Many methods are used to detect the epigenetic status with the combination of sequencing and chemical methods, such as `Me‑DIP` and `oxBS‑seq` for `5mC`; `TAB-seq` and `hme‑DIP` for `5hmC`; `6mA‑DIP` and `6mA‑RE-seq` for `6mA`. Aside from these, third-generation sequencing methods, which refer to direct single-molecule sequencing, such as Oxford Nanopore Sequencing and PacBio SMRT Sequencing, can detect epigenetic status at single-base resolution. BMplot is designed to visualize single-base resolution epigenetic data by considering the strand, motif, and additional information.

BMplot provides two functions to visualize the base modification. `getBaseModificationDf()` gets the information about base modification in each base and organizes it into a data frame. `plotBaseModificationProf()` can return an ggplot object that can be used to visualize the base modification using the data frame returned by `getBaseModificationDf()`.

## `bmData` object preparation

Any regions of user intesets could be visualized in BMplot. With the data provided by users, BMplot creates the `bmData` object to store comprehensive information about base modifications in DNA or RNA. 

DNA methylation is one of the base modifications that plays an important role in various biological processes. As the differentially methylated regions (DMR) are most of the command intesests, here we use the data from WGBS (whole-genome bisufile sequencing) as the showcase to show how to prepare the `bmData` object in BMplot. 

## Base modification information extraction.

For a better experience, we made some example data for users to go through BMplot. Here, we extracted the information from the `bmData` object and mapped it to the sequence data stored in `BSgenome.Athaliana.TAIR.TAIR9` by providing the base and motif of our interest.

```{r eval=FALSE}
library(BSgenome.Athaliana.TAIR.TAIR9)
BSgenome_thaliana <- BSgenome.Athaliana.TAIR.TAIR9

# A_thaliana_dmR and A_thaliana_BSobj are internal sample data
thaliana_df <- getBaseModificationDf(region = A_thaliana_dmR,
                                     BSgenome = BSgenome_thaliana,
                                     input = A_thaliana_BSobj,
                                     base = "C",
                                     motif = c("CG","CHG","CHH"))
```

The data frame has six columns. `coordinate` refers to the coordinates of each base in the region; `motif` refers to the motif of the methylation; `strand` refers to the strand information of the base modification; `sample` refers to the sample name of each sample; `value` refers to the coverage of each base or the methylation level of each position; `type` decides whether the category should be coverage or methylation level.


## Base modification visualization

```{r eval=FALSE}
plotBaseModificationProf(thaliana_df,switch_y_value = T)
```

![](vignettes/figures/readme.png)

## Additional information visualization

### Highlight

Users can use the parameter of highlight_lim to emphasize the regions in `BMplot`.

```{r eval=FALSE}
plotBaseModificationProf(thaliana_df,switch_y_value = T,
                         highlight_lim = c(5651586,5651913))
```

![](vignettes/figures/highlight.png)
This highlight function can also be applied to list regions.
```{r eval=FALSE}
BSgenome_human <- BSgenome.Hsapiens.UCSC.hg19
df_list <- getBaseModificationDf(region = Human_dmR[1:2,],
                                 BSgenome = BSgenome_human,
                                 input = Human_BSobj,
                                 base = "C",
                                 motif = c("CG","CHH","CHG"))

plotBaseModificationProf(df_list,nrow = 1,switch_y_value = T,
                         highlight_lim = list(c(1100262,1100463),
                                              c(1173154,1173300)))
```
![](vignettes/figures/highlight_list.png)

### Gene track

Users can add the gene track in addition to the profile of base modification. The `GeneModel` parameter allows users to plot the gene track of a chosen database. The `ggbio` was used to plot the gene track in accordance with the `GeneModel` parameter. `GeneModel` can be an OrganismDb object, TxDb object, EnsDb object, or GrangeList object.


```{r eval=FALSE}
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(BSgenome.Hsapiens.UCSC.hg19)
BSgenome_human <- BSgenome.Hsapiens.UCSC.hg19
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene

region <- Human_dmR[1,]
region <- region[,c(1:3)]
region[1,c(2,3)] <- c(10000,25000)
human_df <- getBaseModificationDf(region = region,
                                  BSgenome = BSgenome_human,
                                  input = Human_BSobj,
                                  base = "C",
                                  motif = c("CG","CHH","CHG"))

plotBaseModificationProf(human_df,
                         switch_y_value = T,
                         GeneModel = txdb,
                         highlight_lim = c(10497,11000))

```
![](vignettes/figures/genetrack.png)


## Installation

You can install the development version of BMplot from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("YuLab-SMU/BMplot")
```

