---
output:
  md_document:
    variant: gfm
html_preview: false
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# BMplot: Base Modification Visulization

<!-- badges: start -->

<!-- badges: end -->

This package implements functions to visualize the profile of base modification in both DNA and RNA.

## Introduction
Epigenetics plays an important role in Biology in the way of changing genetic information without changing nucleic acids. Histone modification and Methylation are two vital mechanisms in Epigenetics. 

Many methods are used to detect the methylation in Epigenentics with the combination of sequencing and chemical methods. `Me‑DIP` and `oxBS‑seq` for `5mC`. `TAB-seq` and `hme‑DIP` for `5hmC`. `6mA‑DIP` and `6mA‑RE-seq` for `6mA`. These methods can be divided into two categories, including `Antibody enrichment techniques` for base fragments and `Single-base resolution techniques` for single base. BMplot is designed to visualize the `Single-base resolution techniques`.

With the data created by the methods mentioned above, BMplot can map the data into the sequence data stored in the set of BSgenome packages and visualize it with great concern to the `strand` and `motif` information.

## Installation

You can install the development version of BMplot from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("YuLab-SMU/BMplot")
```

